Federation Authentication protocol documentation (#1842)
dedis/popstellar: docs/messageData.md

## Requesting a challenge (federation#challenge_request)
The organizer of one of the two LAOs will request a challenge from his server.
This challenge will be then used by the other server to authenticate itself.

<details>
<summary>
💡 See an example
</summary>

```json5
// ../protocol/examples/messageData/federation_challenge_request/federation_challenge_request.json
{
  "object": "federation",
  "action": "challenge_request",
  "timestamp": 1712854874
}

```
</details>

```json5
// ../protocol/query/method/message/data/dataFederationChallengeRequest.json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/message/data/dataFederationChallengeRequest.json",
  "description": "Sent by an organizer client to its server, to retrieve a challenge object",
  "type": "object",
  "properties": {
    "object": {
      "const": "federation"
    },
    "action": {
      "const": "challenge_request"
    },
    "timestamp": {
      "type": "integer",
      "description": "[Timestamp] of the request",
      "minimum": 0
    }
  },
  "additionalProperties": false,
  "required": [

>     "timestamp"

Reviewer: I believe the `object` and `action` field are missing there.

Reviewer 2: These fields are missing in the file ../protocol/query/method/message/data/dataFederationChallengeRequest.json

The server will provide his corresponding organizer with the requested challenge.
<details>
<summary>
💡 See an example
</summary>

```json5
{
    "object": "federation",
    "action": "challenge",
    "value": "82eadde2a4ba832518b90bb93c8480ee1ae16a91d5efe9281e91e2ec11da03e4",
    "valid_until": 1712854874
}
```
</details>

```json5
// ../protocol/query/method/message/data/dataFederationChallenge.json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/message/data/dataFederationChallenge.json",
  "description": "Challenge object in the context of federation authentication",
  "type": "object",
  "properties": {
    "object": {
      "const": "federation"
    },
    "action": {
      "const": "challenge"
    },
    "value": {
      "type": "string",
      "contentEncoding": "hex",
      "pattern": "^[0-9a-fA-F]{64}$",
      "$comment": "A 32 bytes array encoded in hexadecimal"
    },
    "valid_until": {
      "type": "integer",
      "description": "[Timestamp] of the expiration time",
      "minimum": 0
    }
  },
  "additionalProperties": false,
  "required": [
    "value",

>     "valid_until"

Reviewer: Same as above, I believe the `object` and `action` field are missing there.

Reviewer 2: These fields are also missing in the file ../protocol/query/method/message/data/dataFederationChallenge.json

  ]
}

```

## Expecting a connection (federation#expect)

> With this message the organizer lets the server know to expect a federation_init message from the remote party idetailed in the message body.

Reviewer 3: ```suggestion
With this message the organizer lets the server know to expect a federation_init message from the remote party detailed in the message body.
```

    "public_key": "UvViTxoKsB3XVP_ctkmOKCJpMWb7fCzrcb1XDmhNe7Q=",
    "server_address": "wss://ethz.ch:9000/server",

>     "challenge": {

Reviewer 3: if the point here is to refer to the original challenge message, why not just refer to its message_id ? That would be entirely sufficient to someone having access to that message, seeing that it is a hash of the data and signature.

Reviewer 4: We have to include the whole message here, because the 'original' challenge message (i.e. the one sent from the server to the organizer after a federation_challenge_request is not signed by the organizer). Upon receiving the message from the server, the organizer must extract the challenge data and create a new challenge message (now signed by him---the organizer) and pack that message into the federation_expect message s.t. the server can then again extract the challenge message from the federation_expect (which is now signed by the organizer) and send it back to the remote server as part of the federation result message.

Typing this out i realize this sounds very complicated but I think we discussed this and ultimately decided that  this is the best (and possibly only) way to implement the scheme with the trust boundaries wie defined.

>             "$comment": "public key of the remote organizer"

Reviewer 3: here and in other places you refer to a "remote" organizer, it might be useful to define this once at the beginning of the federation protocol to clarify who is considered "local" and who's considered remote. You might also want to consider renaming this to "inviting" / "joining" , as that might be less ambiguous.

Reviewer 4: I agree that the current wording can at some points be ambiguous, but I fear that if we define it at the start and subsequently remove these adjectives in the later definitions it could be even more ambiguous.

Maybe it would be good to sort of explain at the beginning that the documentation (and protocol definition) is written from the perspective of a specific party (i.e. an organizer Alice) who takes on defined role (i.e. Alice could be the organizer receiving the challenge from the other)?

        },

>         "challenge": {

Reviewer 3: same question as above, why not just refer to the challenge's `message_id`?

          "$ref": "../message.json",

>           "$comment": "message/message containing a FederationChallenge data"

Reviewer 3: what's a `message/message` ? :-)

>           "$comment": "message/message containing a FederationChallenge data"

Reviewer 4: ```suggestion
          "$comment": "message containing a FederationChallenge"
```

Reviewer 2: We are using the general message found at protocol/query/method/message/message.json, where the data field contains a dataFederationChallenge

## Initiating a connection (federation#init)
This message is used to let a server know to initiate a connection with a remote server.

> It contains the necessary connection details, and a challenge which the remote party will expect to receive and the server will use to authenticate himself.

Reviewer 3: `remote party` is ambiguous. Server or client ?

> It contains the necessary connection details, and a challenge which the remote party will expect to receive and the server will use to authenticate himself.

Reviewer 4: ```suggestion
It contains the necessary connection details, and a challenge which the remote server will expect to receive and the server will use to authenticate himself.
```

{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/message/data/dataFederationInit.json",
  "description": "Sent by an organizer client to its server, initiates a connection to a remote LAO",
  "type": "object",
  "properties": {
    "object": {
      "const": "federation"
    },
    "action": {
      "const": "init"
    },
    "lao_id": {
      "type": "string",
      "contentEncoding": "base64",
      "$comment": "ID of the remote LAO"
    },
    "server_address": {
      "type": "string",
      "pattern": "^(ws|wss):\/\/.*(:\\d{0,5})?\/.*$",
      "$comment": "public address of the remote organizer server"
    },
    "public_key": {
      "type": "string",
      "contentEncoding": "base64",
      "$comment": "public key of the remote organizer"
    },
    "challenge": {
      "$ref": "../message.json",
      "$comment": "message/message containing a FederationChallenge data"
    }
  },
  "additionalProperties": false,
  "required": [
    "object",
    "action",
    "lao_id",
    "server_address",
    "public_key",
    "challenge"
  ]
}

```

> ## Establishing or Aborting the connection (federation#result)

Reviewer 4: ```suggestion
## Finalizing or Disregarding the connection attempt (federation#result)
```

> This message is used to know the status of the connection, whether it was a success and a connection between the two LAOs was established or it was a failure.

Reviewer: Which server send this message ?

Reviewer 4: This message is sent by the server who verifies the message. The purpose of the message is to let the other server (i.e. the one that sent the challenge) know if the connection succeeded or not.

Maybe we could describe it more like this?

```suggestion
This message is sent by the server verifying a given challenge to the server who sent the specific challenge. The purpose of this message is to provide the non-verifying server a final status of the connection, whether it was a success and a connection between the two LAOs is now established or there was a failure. (In this case the `reason` field may be used to provide further details)
```

<details>
<summary>
💡 See an example
</summary>

```json5
// ../protocol/examples/messageData/federation_init/federation_init.json
{
  "object": "federation",
  "action": "result",
  "status": "success",
  "public_key": "UvViTxoKsB3XVP_ctkmOKCJpMWb7fCzrcb1XDmhNe7Q=",
  "challenge": {
    "data": "eyJvYmplY3QiOiJmZWRlcmF0aW9uIiwiYWN0aW9uIjoiY2hhbGxlbmdlIiwidmFsdWUiOiJlYmEzZTI0ZWZjZDBiNTNmYTY5OTA4YmFkNWQxY2I2OTlkNzk4MGQ5MzEwOWRhMGIyYmZkNTAzN2MyYzg5ZWUwIiwidGltZXN0YW1wIjoxNzEzMzg1NTY4fQ==",
    "sender": "zXgzQaa_NpUe-v0Zk_4q8k184ohQ5nTQhBDKgncHzq4=",
    "signature": "BILYwYkT5tOBL4rCD7yvhBkhAYqRXOI3ajQ2uJ1gAk-g6nRc38vMMnlHShuNCQ3dQFXYZPn37cCFelhWGjY8Bg==",
    "message_id": "sD_PdryBuOr14_65h8L-e1lzdQpDWxUAngtu1uwqgEI=",
    "witness_signatures": []
  }
}
```
</details>

```json5
// ../protocol/query/method/message/data/dataFederationResult.json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/message/data/dataFederationResult.json",

>   "description": "Sent by an server to a remote server, to inform them about the result of the authentication procedure",

Reviewer: typo: 
```suggestion
  "description": "Sent by a server to a remote server, to inform them about the result of the authentication procedure",
```
